engine.js

```diff
@@ -737,37 +737,91 @@ var accumulateResult = function(ifResult, thenResult) {
-                    error.properties['LAR number'] = loanNumber;
-                    error.properties['Reported State Code'] = fipsState;
-                    error.properties['Reported County Code'] = fipsCounty;
-                    error.properties['Reported Census Tract'] = censusTract;
-                    return [error];
+    HMDAEngine.isValidStateCountyCensusTractCombo = function(hmdaFile) {
+       var currentEngine = this,
+            invalidMSAs = [];
+        return Promise.map(hmdaFile.loanApplicationRegisters, function(element) {
+            return currentEngine.apiGET('isValidCensusCombination', [element.fipsState, element.fipsCounty, element.censusTract])
+            .then(function(response) {
+                var result = resultFromResponse(response);
+                if (result.result) {
+                    if ((element.metroArea === 'NA' && result.msa_code!=='') || element.metroArea!==result.msa_code) {
+                        var error = {'properties': {}};
+                        error.properties['Recommended MSA/MD'] = result.msa_code;
+                        error.properties['LAR number'] = element.loanNumber;
+                        error.properties['Reported State Code'] = element.fipsState;
+                        error.properties['Reported County Code'] = element.fipsCounty;
+                        error.properties['Reported Census Tract'] = element.censusTract;
+                        invalidMSAs.push (error);
+                    }
                 }
+                return Promise.resolve();
+            });
+        }, { concurrency: CONCURRENT_RULES })
+        .then(function() {
+            if (!invalidMSAs.length) {
                 return true;
+            } else {
+                return invalidMSAs;
             }
-            return false;
         });
     };
 
-    HMDAEngine.isNotIndependentMortgageCoOrMBS = function(respondentID, agencyCode) {
-        return this.apiGET('isNotIndependentMortgageCoOrMBS', [agencyCode, respondentID])
-        .then(function(response) {
-            return resultFromResponse(response).result;
-        });
-    };
-
-    HMDAEngine.isMetroAreaOnRespondentPanel = function(metroArea, respondentID, agencyCode) {
-        return this.apiGET('isMetroAreaOnRespondentPanel', [agencyCode, respondentID, metroArea])
-        .then(function(response) {
-            return resultFromResponse(response).result;
+    HMDAEngine.isMetroAreaOnRespondentPanel = function(hmdaFile) {
+        var currentEngine = this,
+            invalidMSAs = [];
+        return Promise.map(hmdaFile.loanApplicationRegisters, function(element) {
+            return currentEngine.apiGET('isNotIndependentMortgageCoOrMBS', [element.agencyCode, element.respondentID])
+            .then(function(response) {
+                var result = resultFromResponse(response);
+                if (result.result) {
+                    var validActionTaken = ['1', '2', '3', '4', '5', '7', '8'];
```

Comment:
I just noticed you pulled this action taken condition out of the rule-spec language and into custom code. What made you decide to do that?